src/Mod/Part/Gui/TaskAttacher.cpp

```diff
@@ -345,47 +350,194 @@ bool TaskAttacher::updatePreview()
+}
+
+void TaskAttacher::processSelection(App::DocumentObject*& rootObj, std::string& sub)
+{
+    // The reference that we store must take into account the hierarchy of geoFeatures. For example:
+    // - Part
+    // - - Cube
+    // - Sketch
+    // if sketch is attached to Cube.Face1 then it must store Part:Cube.Face3 as Sketch is outside of Part.
+    // - Part
+    // - - Cube
+    // - - Sketch
+    // In this example if must store Cube:Face3 because Sketch is inside Part, sibling of Cube.
+    // So placement of Part is already taken into account.
+    // - Part1
+    // - - Part2
+    // - - - Cube
+    // - - Sketch
+    // In this example it must store Part2:Cube.Face3 since Part1 is already taken into account.
+    // - Part1
+    // - - Part2
+    // - - - Cube
+    // - - Part3
+    // - - - Sketch
+    // In this example it's not possible because Sketch has Part3 placement. So it should be rejected
+    // So we need to take the selection object and subname, and process them to get the correct obj/sub based
+    // on attached and attaching objects positions.
+
+    std::vector<std::string> names = Base::Tools::splitSubName(sub);
+    if (!rootObj || names.size() < 2) {
+        return;
+    }
+    names.insert(names.begin(), rootObj->getNameInDocument());
+
+    App::Document* doc = rootObj->getDocument();
+    App::DocumentObject* attachingObj = ViewProvider->getObject(); // Attaching object
+    App::DocumentObject* subObj = rootObj->getSubObject(sub.c_str()); // Object being attached.
+    if (!subObj || subObj == rootObj) {
+        // Case of root object. We don't need to modify it.
+        return;
+    }
+    if (subObj == attachingObj) {
+        //prevent self-referencing
+        rootObj = nullptr;
+        return;
+    }
+
+    // Check if attachingObj is a root object. if so we keep the full path.
+    auto* group = App::GeoFeatureGroupExtension::getGroupOfObject(attachingObj);
+    if (!group) {
+        if (attachingObj->getDocument() != rootObj->getDocument()) {
+            // If it's not in same document then it's not a good selection
+            rootObj = nullptr;
+        }
+        // if it's same document we keep the rootObj and sub unchanged.
+        return;
+    }
+
+    for (size_t i = 0; i < names.size(); ++i) {
+        App::DocumentObject* obj = doc->getObject(names[i].c_str());
+        if (!obj) {
+            Base::Console().TranslatedUserError("TaskAttacher",
+                "Unsuitable selection: '%s' cannot be attached to '%s' from within it's group '%s'.\n",
+                attachingObj->getFullLabel(), subObj->getFullLabel(), group->getFullLabel());
+            rootObj = nullptr;
+            return;
+        }
+
+        // In case the attaching object is in a link to a part.
+        // For instance : 
+        // - Part1
+        // - - LinkToPart2
+        // - - - Cube
+        // - - - Sketch
+        obj = obj->getLinkedObject();
+
+        if (obj == group) {
+            ++i;

```

Comment:
How do you know that `i` was not `names.size()-1` before this call? This chunk of code has a pretty deep code smell. Is there a way to refactor that doesn't rely on using integer indices? It's begging for a buffer overrun, etc.